examples/unadf.c

```diff
@@ -580,17 +579,57 @@ mode_t permissions(struct AdfEntry *e) {
         ((e->access >> 5) & 0070); /* rwx for group */
 }
 
-/* replace all special characters (forbidden on most systems) with an underscore ('_') */
-int replace_not_allowed_chars ( char * const path )
-{
-    const char not_allowed_chars[] = ":*\"?<>|\0";
-    int replace_count = 0;
-    for ( const char *c = not_allowed_chars ; *c ; ++c ) {
-        char *pathc = path;
-        while ( ( pathc = strchr ( pathc, *c ) ) ) {
-            *pathc = '_';
-            replace_count++;
+/* alter filenames that would cause problems on Windows:
+ * https://learn.microsoft.com/en-us/windows/win32/fileio/naming-a-file
+ * 1. replace characters : * " ? < > | / \ and chars 1-31 with underscore
+ * 2. replace any trailng dot or space with underscore (breaks Windows File Explorer)
+ * 3. if filename is CON PRN AUX NUL COM[1-9] LPT[1-9], regardless of case or
+ *    extension, insert an underscore (vestiges of CP/M retained by Windows)
+ *
+ * In this last case, the name will be _extended_ by 1 character!
+ */
+void fix_win32_filename(char *name) {
+    /* work forwards through name, replacing forbidden characters */
+    char *o;
+    for (o = name; *o; o++) {
+        if (*o < 32   || *o == ':' || *o == '*' || *o == '"' || *o == '?' ||
+            *o == '<' || *o == '>' || *o == '|' || *o == '/' || *o == '\\')
+        {
+            *o = '_';
         }
     }
-    return replace_count;
+    /* work backwards through name, replacing forbidden trailing chars */
+    for (o--; o >= name && (*o == '.' || *o == ' '); o--) {
+       *o = '_';
+    }
+
+    /* temporarily remove file extension */
+    char *ext = strchr(name, '.');
+    if (ext) {
+        *ext = 0;
+    }
+
+    /* is the name (excluding extension) a reserved name in Windows? */
+    const char *reserved_names[] = {
+        "CON", "PRN", "AUX", "NUL",
+        "COM1", "COM2", "COM3", "COM4", "COM5", "COM6", "COM7", "COM8", "COM9",
+        "LPT1", "LPT2", "LPT3", "LPT4", "LPT5", "LPT6", "LPT7", "LPT8", "LPT9"
+    };
+    for (int i = 0; i < sizeof(reserved_names)/sizeof(char **); i++) {
+       if (strcasecmp(reserved_names[i], name) == 0) {
+           if (ext) {
+               /* restore extension and insert underscore, PRN.txt -> PRN_.txt */
```

Comment:
if the filename becomes longer, it might write outside of bounds of the previous string, for example if it has been allocated with strdup.

Reply:
The memory allocated for the name is meant to account for that, but it didn't. I've added a fix.